hide show hide if user is not signed in

abeddow91 opened the pull request:
## What does this change?
This update places the show/hide container toggles behind a sign-in requirement for users interacting with beta containers. 

In the immediate, the behaviour is as follows:

- If the container is a beta container, we only allow show hide if the user is signed in
- If the container is a existing / legacy container, we always allow show hide.

Following the launch of the new containers, this feature will be extended to all containers (ie users must be signed in to use show/hide on any container), ensuring consistent functionality across the platform. A tracking ticket has been created for this follow-up work.

The feature operates as follows: each show/hide button is associated with a beta container data flag. If the flag is set to true and the user is not signed in, the button is hidden from view. Since this logic resides within a useEffect hook dependent on the user's sign-in status, signing in will dynamically make the buttons visible again. As the code needs to know if the user is signed in or not, the code has an early return when the sign in status is "Pending" to prevent any incorrect renders of the toggle. 

Additionally, this PR includes a refactor of the show/hide island and adds comments with the aim to improve readability of maintainability of the island. 

This PR is dependent on https://github.com/guardian/dotcom-rendering/pull/13162, otherwise - as seen in the screenshots - the scrollable nav buttons and the show / hide controls overlap.

## Why?
This change supports the goal of restricting the show/hide functionality to signed-in users.

## Screenshots

| Exisiting containers & not signed in       |  Beta containers & not signed in           |  Beta containers & signed in          |
| ----------- | ---------- | ---------- |
| ![before][] | ![after][] | ![after1][] |


[before]: https://github.com/user-attachments/assets/0f9d6fe5-fbdf-412a-ac24-28f3242354e9
[after]: https://github.com/user-attachments/assets/0cb68b35-9431-4ad1-bc49-73bb276aaeec
[after1]: https://github.com/user-attachments/assets/05625d75-8850-4b01-bf3b-a45814321369


<!--
You can add extra rows by repeating the last row in the table and then using new unique labels. E.g.

| ![before2][] | ![after2][] |

You can then reference the labels and map them to corresponding links.

[before2]: https://example.com/before2.png
[after2]: https://example.com/after2.png
-->

<!--
## Running Chromatic

In order to run Chromatic as part of the CI checks, you will need to add the `run_chromatic` label to your PR. Once the label is added Chromatic will run on every push.

Please only add this once you are ready to check for visual regressions, our intention here is to reduce the amount of time Chromatic is run without being looked at.
-->

<!--
## Unexplained Chromatic diffs

We use Chromatic for visual regression testing on our Storybook stories. It's
generally pretty good, but it sometimes gives 'false positives' -- it seems to
detect a change in a component which hasn't changed, or which hasn't been
affected by the code in your PR.

If you've looked at the Chromatic diffs and can't see any connection to your
code, please reach out to a member of the Web Experiences team, who will be able
to advise. It would also be helpful to add the false positive to our
[ongoing log of false positives](https://docs.google.com/spreadsheets/d/1FvItNTMFXIpI4rCrZ4mQ0CRouT06sSVro168f6oKPm4/edit?usp=drive_open&ouid=117150399571694275917#gid=0).
-->

domlander commented on dotcom-rendering/src/components/ShowHideContainers.importable.tsx:
```diff
@@ -53,15 +73,22 @@ export const ShowHideContainers = () => {
 			if (!sectionId) continue;
 
-			e.onclick = () => toggleContainer(sectionId, e);
+			if (isSignedIn === false && isBetaContainer === 'true') {
```
Following on from my [other comment](https://github.com/guardian/dotcom-rendering/pull/13176#issuecomment-2607193951), if we are later going to remove the `isBetaContainer` check, I would be tempted to separate out the logic for what happens when the container is a beta container from when it isn't, i.e. favouring duplication to make it easier to remove the unnecessary code later on.

i.e.
```
if (isBetaContainer === 'true') {
...
} else {
...
}
```

abeddow91 replied:
I have reworked this logic so (hopefully) its easier to understand. 

In the immediate, the behaviour is as follows: 
 - If the container is a beta container, we only allow show hide if the user is signed in
 - If the container is a existing / legacy container, we always allow show hide.

abeddow91 commented:
> Are users that are not signed in on non-beta containers allowed to see the show/hide buttons? It appears that they can but the hidden state is not saved on refreshing the page when I tested locally.

Thanks for checking, there was a bug in the initial pull where i wasn't accounting for a "pending" signed in state which was causing mulitple renders and as such we were losing the correct collapsed state of containers. As we depend on the sign in status to decide if we show the toggle or not, I have added an early return from the loop if the sign in status is pending. 

I have also reworked this logic so (hopefully) its easier to reason about.

In the immediate, the behaviour is as follows: 
 - If the container is a beta container, we only allow show hide if the user is signed in
 - If the container is a existing / legacy container, we always allow show hide.
 
 Following the launch of the new containers, all containers will adopt the show / hide behind sign in flag behaviour.

domlander commented on dotcom-rendering/src/components/ShowHideContainers.importable.tsx:
```diff
@@ -18,58 +25,99 @@ const getContainerStates = (): ContainerStates => {
-		};
+		for (const button of showHideButtons) {
+			/**
+			 * We need if a user is signed in before we can make any further decisions about show/hide buttons.
```
```suggestion
			 * We need to know if a user is signed in before we can make any further decisions about show/hide buttons.
```
